implement `__traits(getCppNamespaces, symbol)`

thewilsonator opened the pull request:
Spec PR dlang/dlang.org#2867

Adds `__traits(getCppNamespaces, symbol)` which retrieves, as a tuple
of strings, the C++ namespaces a symbol resides in.

This enables reflection and compile time code generation capabilities
to C++ namespaces.

thewilsonator commented on src/dmd/traits.d:
```diff
@@ -1945,6 +1945,39 @@ Expression semanticTraits(TraitsExp e, Scope* sc)
+        import dmd.asttypename;
+        writeln("ASTtypeName ",astTypeName(oexp), " ", oexp.toString());
+        TupleExp tup;
+        if (auto otup = oexp.isTupleExp())
```
This seems to not be triggered by the test 
```
extern(C++, mixin("AliasSeq!(`ns`, `nt`)")) struct GetNamespaceTest3 {}
```
not sure how to trigger.

BorisCarvajal replied:
I just checked it seems only the last element get assigned to cppnamespace.exp not a tuple.
In `extern(C++, mixin("AliasSeq!(`ns`, `nt`)")) struct GetNamespaceTest3 {}` only the StringExp "nt" is there.

thewilsonator replied:
Yes see the loop in the else branch. @Geod24 said this was because we are post semantic here.

thewilsonator replied:
https://github.com/dlang/dmd/blob/master/src/dmd/attrib.d#L502-L511

BorisCarvajal replied:
Yeah sorry I should've read the doc.

thewilsonator replied:
Don't worry, I read it and still got confused by it.

BorisCarvajal replied:
No sure how a tuple could be here, you could run expression semantic first to be really sure and simply delete this branch.
Moreover, semantic is run for each element anyway after the trait tuple creation.

thewilsonator commented on src/dmd/traits.d:
```diff
@@ -1945,6 +1945,39 @@ Expression semanticTraits(TraitsExp e, Scope* sc)
+            return tup.expressionSemantic(sc);
+        }
+        auto oexp = s.cppnamespace.exp;
+        import std.stdio;
+        import dmd.asttypename;
+        writeln("ASTtypeName ",astTypeName(oexp), " ", oexp.toString());
```
Whoops will remove soon

Geod24 commented:
What's the intent here ? I'm a bit worried, because a string namespace is **not** a symbol that contains other symbol, unlike identifiers namespaces, so this seems flawed.

thewilsonator commented:
Principally to do 
```template ScopeClass(C)
//  if (is(C == class))
{
    static if (__traits(getLinkage, C) == "C++")
    {
        extern(C++, __traits(getNamespaces, symbol)) // <<< Here 
        {
            extern(C++, class)
            pragma(mangle, C.stringof)
            struct ScopeClass
            {
                char[__traits(classInstanceSize, C)] buffer;
                //... all the things ...
            }
        }
    }
}
``` 
for https://github.com/dlang/dmd/pull/11835 to enable setting the namespaces for symbols. LMK if I need to explain in more depth. also cc @TurkeyMan

WalterBright reviewed (dismissed):
Complete absence of any sort of documentation or rationale. This is not acceptable for core language changes.

thewilsonator commented:
Spec PR https://github.com/dlang/dlang.org/pull/2867

thewilsonator commented:
(Modulo network errors) this should be ready to go.

thewilsonator commented:
@WalterBright spec PR an changelog added.

Geod24 reviewed (changes requested):
Good addition, simple and makes sense.
Added a suggestion for the changelog, and I think the core of the code can be greatly simplified. Also need some forward reference tests, as the C++ compatibility layer has suffered from so many bugs in that regard.

thewilsonator commented on changelog/traits-getcppnamespaces.dd:
```diff
@@ -0,0 +1,18 @@
+Add `__traits(getCppNamespaces, symbol)` to retrieve the C++ namespaces a symbol resides in.
+
+Add `__traits(getCppNamespaces, symbol)` which retrieves, as a tuple of strings,
+the the C++ namespaces a symbol resides in.
+
+This enables determining what namespaces a given symbol resides in for use in reflection.
+This can then be used in compile time code generation to control the namespace of generated symbols.
```
It does work with non-string namespaces, as much as id love to get rid of them.

Geod24 replied:
What I mean is that you cannot generate a non-string namespace using the traits. There's no way I can think of to use this trait as an input to `extern(C++, ns)` and have `ns` be non-string namespace.

thewilsonator committed:
implement `__traits(getCppNamespaces, symbol)`

Adds `__traits(getCppNamespaces, symbol)` which retrieves, as a tuple
of strings, the C++ namespaces a symbol resides in.

This enables reflection and compile time code generation capabilities
to C++ namespaces

thewilsonator commented:
@Geod24 all comments addressed.

Geod24 commented on src/dmd/traits.d:
```diff
@@ -1945,6 +1945,42 @@ Expression semanticTraits(TraitsExp e, Scope* sc)
+            // and `b`'s points to `c`. Our entry point is `a`.
+            for (auto ns = p.cppnamespace; ns !is null; ns = ns.cppnamespace)
+            {
+                ns.dsymbolSemantic(sc);
```
Is this needed ? After semantic has been run on a symbol, the namespaces should be fully expanded.

thewilsonator replied:
I have no idea, but it worked!

thewilsonator replied:
Note that this fixed the forward declaration issues.

Geod24 commented on src/dmd/traits.d:
```diff
@@ -1945,6 +1945,42 @@ Expression semanticTraits(TraitsExp e, Scope* sc)
+                exps.insert(0, ns.exp);
+            }
+        }
+        if (auto d = s.isDeclaration())
+            d.inuse = 0;
```
This and the previous usage of `inuse` seems a bit dangerous. I'll give it a test tonight.

thewilsonator replied:
Thats how circular reference checking is done, right?

Geod24 replied:
Almost, but usually setting the variable is done around a semantic call, while here there seem to be multiple semantic calls involved.

thewilsonator commented:
@Geod24 any further comments?

Geod24 commented:
I wanted to give this a brief test w.r.t. forward reference. Don't know if I can get to it soon tho. So let's add The Label.

thewilsonator commented:
3 days are up

Geod24 commented:
And my tests didn't find anything, so let's merge!
